### DISCUSSION TITLE
change the default value of `ignore_trivial` param in stump

### DISCUSSION THREAD

**Discussion by NimaSarajpoor** [TDAmeritrade/stumpy#691] [Ideas]
I was wondering if it would be reasonable to set `ignore_trivial` to default to `None`. Then, if user does not set it to `True` or `False`, we can set it automatically:
```
if ignore_trivial is None:
    if T_B is None:
          ignore_trivial = True
    else: 
          ignore_trivial = False
```

Because, I made a mistake once when I was trying to calculate AB-join. I then realized that I should have set `ignore_trivial` param to False.

**Comment by seanlaw** [comment]
It's probably fine. I'm just trying to think whether there are any situations where `T_B is not None` and `ignore_trivial` should be `True`?

**Reply by NimaSarajpoor** [reply]
I can think of one example. However, I am not sure if it is going to be useful in real-world problems.

Let's say I have a timeseries `T` and I am interested to consider subsequences in `T[start:stop]` and find their NN in the whole  `T`. 

Let's see how one can achieve this. I can think of three different solutions:

* Calculate the full matrix profile on T, and then just consider `mp[start : stop - m + 1]`
* use `for-loop`, and in each iteration, calculate distance profile `stumpy.core.mass(Q, T)`, where `Q` is a subsequence in `T[start : stop]`
* Or, we can use mp-join

However, the last approach is tricky. Because, we cannot say the answer is `stumpy.stump(T[start:stop], m, T, ignore_trivial=False)`. Because, the `T_A` is coming from `T_B`. Hence, we need to ignore trivial neighbors. we may do:

```
T_new = T.copy()
T_new[:start] = np.nan
T_new[stop:] = np.nan

mp = stumpy.stump(T_new, m, T, ignore_trivial=True) # NOTE THAT `ignore_trivial` is set to `True`
```

----

**One last note:**
Still, there is one (small?) challenge. I think the computation time of this final approach should be close to the very first approach, i.e. `calculate the full matrix profile` unless we modify `stumpy.stump` to skip the first nan values and the last nan values.
(In the current  implementation of `stump`, we do not skip this! We calculate `cov` right from the beginning. However, in this case, we just need to skip the nan values till we reach the very first non-nan value and start from there)

**Reply by seanlaw** [reply]
> I was trying to calculate AB-join. I then realized that I should have set ignore_trivial param to False.

What if we went a different direction and, instead, what if we simply printed out a warning/suggestion when we detect that:

1. `ignore_trivial == True`
2. `T_A` != `T_B`

So, something like:

```
# In core.py, we add:
def check_memory_overlap(a, b):
    return np.shares_memory(a, b)

# In `stumpy.stump`, we add the following before `preprocess` or `_preprocess` is called!
if ignore_trivial and check_memory_overlap(T_A, T_B) is False:
    logger.warning(f"`ignore_trivial == True` and the input arrays share overlapping memory space")
    logger.warning(f"This is exceptionally rare and you may want to try setting `ignore_trivial=False` for AB-joins")
```

Thus, this warning would be NOT be triggered if you did:

```
T = np.random.rand(1000)
m = 50
mp = stumpy.stump(T, m, T)
```

and it would also NOT trigger a warning with:

```
T_A = np.random.rand(1000)
T_B = T_A
m = 50
mp = stumpy.stump(T_A, m, T_B)
```

but a warning would be triggered with:

```
T_A = np.random.rand(1000)
T_B = T_A.copy()
m = 50
mp = stumpy.stump(T_A, m, T_B)
```

and also triggered with:

```
T_A = np.random.rand(1000)
T_B = np.random.rand(1000)
m = 50
mp = stumpy.stump(T_A, m, T_B)
```

This maintains backwards compatibility while still accomplishing the task of warning the user

**Reply by NimaSarajpoor** [reply]
> This maintains backwards compatibility

I assume you are thinking about a user who might have done something like this: `stumpy.stump(T_A, m, T_B)`, and they intentionally considered the default `True` in this case. So, if we change API to `stump(..., ignore_trivial=None, ...)`, it might change their output if we automatically set it to `False` in `AB-join`. 

---

I can see your point. I was trying to see if we can get rid of this warning in `stumpy/stump.py`:

```
# line 581

if ignore_trivial and core.are_arrays_equal(T_A, T_B) is False:  # pragma: no cover
        logger.warning("Arrays T_A, T_B are not equal, which implies an AB-join.")
        logger.warning("Try setting `ignore_trivial = False`.")

```

Because, I just thought, in most cases, when users provide `T_B`, they probably want to also do `ignore_trivial=False`, and they know what they are doing. So, that is why I thought it would be cooler(?) to make life easier for user. 

On the other hand, I can see your concern for the backwards compatibility. So, if you prefer to keep the current default setting, then I think the current warning message should be fine. 

Nevertheless, if you want to improve `warning`, how about the following warning condition / message:

```
if ignore_trivial==False and check_memory_overlap(T_A, T_B):
    logger.warning(f"`ignore_trivial == False` but the input arrays share overlapping memory space")
    logger.warning(f"This is exceptionally rare and you may want to try setting `ignore_trivial=True` for AB-joins in this particular case")
```
I think this one somehow complements the already-existing warning in `stumpy/stump.py` shown above. I think this is what you had in mind (?)

**Reply by seanlaw** [reply]
> I was trying to calculate AB-join. I then realized that I should have set ignore_trivial param to False.

So what happened when you did this? Did you encounter the warning? I totally forgot that this warning existed already

**Reply by NimaSarajpoor** [reply]
> So what happened when you did this? Did you encounter the warning? I totally forgot that this warning existed already

I thought I did.. but I just checked and I got no warning. 
```
T_A = np.array([1, 2, 3, 0, 1, 2, 4, 0, 1, 2, 3.0, -4, -10, 1])
T_B = np.array([11, 12, 31, 10, 11, 12, 41, 10, 11.0, 21, 31.0, -41, -101, 11])
stumpy.stump(T_A, 3, T_B)

# I got output with no warning
```

Maybe that is why I remembered this issue! (This is not a recent experience. I have been working on Novelets and I had to compute `mp-join` and I remembered that I had an experience where something was off about default of `ignore_trivial` param and I decided to share it with you)

---

Also, I just noticed that, in the beginning of the function `stump`, we have the following lines of code:
```
if T_B is None:
        T_B = T_A
        ignore_trivial = True
```
Therefore, we overwrite `ignore_trivial` when `T_B is None`.  Right? Having said that, I understand that `ignore_trivial=False` can become meaningless in self-join as the minimum distance  will be 0.0 (self-match). So, the problem is when T_B is not None, and in that case, I believe we should have set it to False unless it is specified by user.

**Reply by NimaSarajpoor** [reply]
@seanlaw 
Apparently `is` checks for the `id`, and it was very interesting when I realized two `False` can have different `id`s !!!

**Example 1**
```
T_A = np.array([1, 2, 3, 0, 1, 2, 4, 0, 1, 2, 3.0, -4, -10, 1])
T_B = np.array([11, 12, 31, 10, 11, 12, 41, 10, 11.0, 21, 31.0, -41, -101, 11])
out = stumpy.core.are_arrays_equal(T_A, T_B)
>>> print(out) 
False

>>> id(out)
140733916434904

>>>id(False)
140734028742000

x = out is False
>>> x
False
```

**Example 2**
```
out = False
x = out is False
>>> x 
True
```

So, we just need to replace `is` with `==`.

**Reply by seanlaw** [reply]
Ouch! I guess that check NEVER actually worked and it looks like we have this in multiple places when I check `grep "is False" *py`. 🤦‍♂️  Sorry about that!

As you've proposed, I think we should:

1. Replace `is` with `==`
2. Refactor: Define a new function in `core.py` to `check_ignore_trivial`

Something like:

```
# core.py
def check_ignore_trivial(T_A, T_B, ignore_trivial):
    if ignore_trivial == False and core.are_arrays_equal(T_A, T_B):  # pragma: no cover
        logger.warning("Arrays T_A, T_B are equal, which implies a self-join.")
        logger.warning("Try setting `ignore_trivial = True`.")

    if ignore_trivial and core.are_arrays_equal(T_A, T_B) == False:  # pragma: no cover
        logger.warning("Arrays T_A, T_B are not equal, which implies an AB-join.")
        logger.warning("Try setting `ignore_trivial = False`.")
```

Or some reasonable combination of warnings that we described above. Then, finally, we add some unit tests to test this function to ensure that it does what we would expect.

What do you think? Personally, as annoying as it may be, I don't love `ignore_trivial=None` because we are now committing ourselves to inferring what the user wants. In most cases, we'll likely be okay but I am genuinely concerned about the rare edge cases where we would infer incorrectly. This is the whole "shooting yourself in the foot" that I am thinking of but maybe I am overthinking it. Again, I get annoyed by it too and, in the end, we should target the majority (obvious) use case so I am open to it.

**Reply by NimaSarajpoor** [reply]
Thanks for catching the other cases via `grep`. Yes... refactoring would be great!!! 

> Personally, as annoying as it may be, I don't love `ignore_trivial=None` because we are now committing ourselves to inferring what the user wants. In most cases, we'll likely be okay but I am genuinely concerned about the rare edge cases where we would infer incorrectly.

You are right... I have been thinking about it lately and I think the `warning` is better. Although this warning might annoy the person who want to intentionally set `ignore_trivial=True` in `mp-join` (because they are going to get warning again and again), I think it is safer  and, as you once said, it somehow forces the user to THINK about what they are doing. Also, the case <`ignore_trivial=True` in `mp-join`> is rare and thus, that annoying thing is okay.

**Reply by seanlaw** [reply]
> they are going to get warning again and again

Technically, they can turn that specific warning off if they don't care about it. There are straightforward ways to do this in Python

>  it somehow forces the user to THINK about what they are doing

Exactly! The warning is a gentle nudge to let the user know "hey, it looks like you are doing something that is atypical/rare so please take care but proceed carefully if you know what you are doing"

**Reply by NimaSarajpoor** [reply]
> Technically, they can turn that specific warning off if they don't care about it. There are straightforward ways to do this in Python

My bad! I totally forgot about that! 

---

I am going to open an issue shortly.

**Comment by seanlaw** [comment]
Tests should probably look something like:

```
def test_check_ignore_trivial():
    T_A = np.random.rand(10)
    T_B = np.random.rand(20)
    with warnings.catch_warnings(record=True) as w:
        ignore_trivial = False
        core.check_ignore_trivial(T_A, T_A, ignore_trivial)
        assert issubclass(w[-1].category.UserWarning)
        ignore_trivial = False
        core.check_ignore_trivial(T_A, T_B, ignore_trivial)
        assert issubclass(w[-1].category.UserWarning)

        assert "Arrays T_A and T_B" in str(w[-1].message)
```

https://stackoverflow.com/questions/3892218/how-to-test-with-pythons-unittest-that-a-warning-has-been-thrown

This will need to be added after the GPU stuff is cleared up
